### PR TITLE
Add type hints to codebase

### PR DESCRIPTION
**Pull request by MasterOdin**
This PR will add type hints to all possible places in the codebase. I debated on adding `declare(strict_types=1);` at the same time, but not sure if should be done at the same time, or separately. Putting this up in draft form to keep track of progress, and hopefully help ensure no one else spends time on doing this concurrently.

### CODE REVIEW

**Comment by dereuromark** [comment]
I would hold off on the strict_types part, as this can create some more pitfalls where currently auto-casting works (to not break more than necessary).

File: src/Phinx/Db/Table/Index.php
Comment (MasterOdin) [inline src/Phinx/Db/Table/Index.php:67]: This change makes this method equivalent to how `setColumns` works in the `ForeignKey` class.

File: src/Phinx/Db/Table/Table.php
Comment (MasterOdin) [inline src/Phinx/Db/Table/Table.php:82]: The other classes under this namespace all return their class instance at the conclusion of `setOptions`, so adding this for consistency.

File: src/Phinx/Db/Action/ChangeColumn.php
Comment (dereuromark) [inline src/Phinx/Db/Action/ChangeColumn.php:55]: I wonder if we can check more of those "mixed" ones and fix up the docblock to be more specific.

File: src/Phinx/Wrapper/TextWrapper.php
Comment (dereuromark) [inline src/Phinx/Wrapper/TextWrapper.php:153]: Like here, too

File: src/Phinx/Db/Table/Column.php
Comment (dereuromark) [inline src/Phinx/Db/Table/Column.php:79]: Thats not necessary, no value is null.

Comment (MasterOdin) [reply src/Phinx/Db/Table/Column.php:79]: Alright, was not sure as some parameters had an explicit `null` set versus others which just had the typedoc. Can remove them throughout the codebase?

File: src/Phinx/Config/Config.php
Comment (MasterOdin) [inline src/Phinx/Config/Config.php:63]: Weakened this to `\Phinx\Config\ConfigInterface` as I believe if someone extends off `Config` and runs this function, then it would return an instance of the extended class and not `Config` via the usage of the `static` keyword on initializing the class.

File: src/Phinx/Db/Action/AddColumn.php
Comment (MasterOdin) [inline src/Phinx/Db/Action/AddColumn.php:41]: phpcs did not like it having the return type annotation matching the type hint. However, I think both might be wrong here given the static binding, not sure how to resolve? replace `@return self` with `@return static` and remove the inline type hint until phinx exclusively uses PHP 8.0? This is the same with other files in the `Action` sub-folder.

Though could maybe just remove the usage of `static` here and use `new AddColumn` and not have this class be extendable given it doesn't use an interface.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/AdapterInterface.php:444]: This was changed from `int` to `float` to accommodate the `MysqlAdapter::INT_BIG` constant which has a value of `18446744073709551615` where that is treated as a `float` within PHP. Typing this (or having the integer typedoc) is then technically incorrect, and would throw a type error if we were to turn on `declare(strict_types=1);` and then `MysqlAdapter::INT_BIG` was used. The best option perhaps would be to type as `float`, and then just document that while it's a `float`, only integer like numbers should be used and decimal values will not work. The problem though is that now integers passed into the function are implicitly converted to floats (e.g. `4` -> `4.0`) which is a bit odd.

Could also maybe insert a:
```
if ($limit <= PHP_INT_MAX) {
    $limit = (int)$limit;
}
```
in the implemented functions to retain the current behavior best we can while having proper typehints?

File: src/Phinx/Db/Action/ChangeColumn.php
Comment (MasterOdin) [reply src/Phinx/Db/Action/ChangeColumn.php]: Fixed a couple of these. We can probably resolve a number of the `mixed` values throughout the codebase, but I'd rather not bloat this PR more than it already is and treat it more in the spirit of "mechanical replacement" and then do targeted replacements of the remaining mixed values.

File: src/Phinx/Db/Table/Index.php
Comment (MasterOdin) [reply src/Phinx/Db/Table/Index.php:67]: Though the handling of the `$columns` value is very inconsistent across the codebase. A number of the `AlterInstruction` functions within the database adapters accepts a `string $column` property while some accept a `Column $column`, and then elsewhere we have `string|string[] $columns` or just `string[] $columns`. A follow-up PR would perhaps be good to work on making this more consistent throughout.

File: src/Phinx/Db/Table/Column.php
Comment (dereuromark) [reply src/Phinx/Db/Table/Column.php:79]: The code sniffer (at least mine does^^) should remove the overhead `= null` usually. 
We might want to add e.g. https://github.com/spryker/code-sniffer/blob/master/Spryker/Sniffs/Classes/PropertyDefaultValueSniff.php on top

File: src/Phinx/Db/Action/AddColumn.php
Comment (dereuromark) [reply src/Phinx/Db/Action/AddColumn.php:41]: Return types should never be self/static until we are fully PHP8, just omit them, then the docs here are easy enough.
The sniffer usually should check that no other return type is inside the scope of the method, and then a return type is not needed and superfluous.

Comment (MasterOdin) [reply src/Phinx/Db/Action/AddColumn.php:41]: Not sure I follow on what you want to see for this specific function or elsewhere. For any case where the return docstring is `@return self` or `@return static`, just omit the inline return type altogether?

Comment (dereuromark) [reply src/Phinx/Db/Action/AddColumn.php:41]: The functional return type, yes. No return types for self or static references, as they can be purely CS checked.

File: src/Phinx/Config/Config.php
Comment (othercorey) [inline src/Phinx/Config/Config.php:369]: Add a `use` statement for this.

File: src/Phinx/Config/ConfigInterface.php
Comment (othercorey) [inline src/Phinx/Config/ConfigInterface.php:111]: Add a `use` statement for this

File: src/Phinx/Console/Command/AbstractCommand.php
Comment (othercorey) [inline src/Phinx/Console/Command/AbstractCommand.php:47]: Can you actually have a null config?

File: src/Phinx/Db/Action/AddColumn.php
Comment (othercorey) [reply src/Phinx/Db/Action/AddColumn.php:41]: The doc return can can be `static` here. `self` isn't correct for `new static`.

File: src/Phinx/Db/Action/AddForeignKey.php
Comment (othercorey) [inline src/Phinx/Db/Action/AddForeignKey.php:44]: Use `static` type here too

File: src/Phinx/Db/Action/AddForeignKey.php
Comment (othercorey) [inline src/Phinx/Db/Action/AddForeignKey.php:46]: Can omit the return type here I think.

File: src/Phinx/Db/Action/AddIndex.php
Comment (othercorey) [inline src/Phinx/Db/Action/AddIndex.php:41]: `static` here too

File: src/Phinx/Db/Adapter/AdapterFactory.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterFactory.php:64]: This description should be updated.

File: src/Phinx/Db/Adapter/AdapterFactory.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterFactory.php:143]:
```suggestion
     * @return object|string
```

File: src/Phinx/Db/Adapter/AdapterFactory.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterFactory.php:121]:
```suggestion
     * @param object|string $class Class
```

File: src/Phinx/Db/Adapter/AdapterFactory.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterFactory.php:67]:
```suggestion
     * @param object|string $class Class
```

File: src/Phinx/Db/Adapter/AdapterFactory.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterFactory.php:55]:
```suggestion
     * @var object[]|string[]
```

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:122]: The function should not have a return type for $this hints.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:137]: Remove return type hint for `$this`.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:165]: Same here.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:173]: And here.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:188]: Same here. Remove return type hint for $this

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:196]: Same here. Remove return type hint for $this

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:278]: Add a `use` statement for this.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (othercorey) [inline src/Phinx/Db/Adapter/AdapterInterface.php:440]:
```suggestion
     * @param \Phinx\Util\Literal|string $type Type
```

File: src/Phinx/Db/Table/Column.php
Comment (othercorey) [inline src/Phinx/Db/Table/Column.php:89]: There's no need to default to null for any of these.

File: src/Phinx/Db/Table/Index.php
Comment (othercorey) [inline src/Phinx/Db/Table/Index.php:42]: No need to default to null

File: src/Phinx/Migration/AbstractMigration.php
Comment (othercorey) [inline src/Phinx/Migration/AbstractMigration.php:41]: No need to default to null

File: src/Phinx/Migration/AbstractMigration.php
Comment (othercorey) [inline src/Phinx/Migration/AbstractMigration.php:212]: Add a `use` statement.

File: src/Phinx/Migration/Manager/Environment.php
Comment (othercorey) [inline src/Phinx/Migration/Manager/Environment.php:34]: No need to default to null

File: src/Phinx/Migration/MigrationInterface.php
Comment (othercorey) [inline src/Phinx/Migration/MigrationInterface.php:62]: Remove return type for `$this` hints

File: src/Phinx/Seed/SeedInterface.php
Comment (othercorey) [inline src/Phinx/Seed/SeedInterface.php:51]: Remove return type for `$this` hints

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php
Comment (othercorey) [inline tests/Phinx/Db/Adapter/MysqlAdapterTest.php:1061]: This feels odd.

**Review by othercorey — CHANGES_REQUESTED** [review]
Reviewed syntax. Left a couple open questions about valid types, but just from a logic point of view.

**Comment by dereuromark** [comment]
Shall we continue/finish this for the next branch?

File: src/Phinx/Console/Command/AbstractCommand.php
Comment (MasterOdin) [reply src/Phinx/Console/Command/AbstractCommand.php:47]: As the config is initialized in its own function, yes. It's why this was necessary for a period of time:
```php
        /** @var \Phinx\Config\ConfigInterface|null $config */
        $config = $this->getConfig();
        if (!$config) {
            $this->loadConfig($input, $output);
        }
```

to make phpstan happy as `$this->getConfig()` was annotated as always returning `ConfigInterface`, but on first call to `bootstrap()` (which holds the above reference), it would be null.

**Comment by MasterOdin** [comment]
@othercorey requested changes have been made.

@dereuromark the failing phpstan test is due to a bug in phpstan: https://github.com/phpstan/phpstan/issues/5369. Could maybe pin the used version to an older one that doesn't have this bug?

File: src/Phinx/Console/Command/Create.php
Comment (dereuromark) [inline src/Phinx/Console/Command/Create.php:204]: the variable is missing here

Comment (MasterOdin) [reply src/Phinx/Console/Command/Create.php]: Added.

The variable isn't strictly necessary as the IDE / phpstan will just apply whatever is there then to the variable in the next line, though having the variable is clearer.

**Comment by dereuromark** [comment]
@othercorey What do you think about the float thingie? Should we keep as is, or do we want to internally work with integers here (as one intuitively would)?

**Comment by othercorey** [comment]
Looks like `limit` is the same as `length` in cakephp. We cast length to int so doesn't seem like it would hurt to do that here. I don't think there are any fractional length types.

Is it possible to cast the limit when reflecting?

**Comment by MasterOdin** [comment]
For the MySQL adapter, the following values are problematic if declaring a `int` limit:

https://github.com/cakephp/phinx/blob/26adc419257b84647602ec4e39e66b850116430b/src/Phinx/Db/Adapter/MysqlAdapter.php#L61

https://github.com/cakephp/phinx/blob/26adc419257b84647602ec4e39e66b850116430b/src/Phinx/Db/Adapter/MysqlAdapter.php#L68

https://github.com/cakephp/phinx/blob/26adc419257b84647602ec4e39e66b850116430b/src/Phinx/Db/Adapter/MysqlAdapter.php#L73-L74

These values all exceed the value of `INT_MAX` on 32-bit systems and `INT_BIG` exceeds `INT_MAX` on 64-bit systems. While these values are set to match the max size of the respective column types, given that these values exist purely as a shorthand to get say `bigint(20)` when doing `->column('foo', 'integer', ['limit' => MysqlAdapter::INT_BIG]`, could just arbitrarily cap them to be values less than `INT_MAX`, and then can declare the `limit` clause as an integer again.

This would be BC breaking, but in a way that I'm not sure anyone would notice. The specifics here are that:
* The values of the above constants would be different
* Assuming the new value of `TEXT_LONG` is now `2147483647` instead of `4294967295`, then values between `2147483647` and `4294967295` would be considered `TEXT_LONG` instead of `TEXT_MEDIUM` (this would apply to the other types in similar way)

However, again, given that these are supposed to be "magic constants" as it were and not to directly depend on the underlying value, should be fine to change?

**Comment by othercorey** [comment]
> These values all exceed the value of `INT_MAX` on 32-bit systems and `INT_BIG` exceeds `INT_MAX` on 64-bit systems. While these values are set to match the max size of the respective column types, given that these values exist purely as a shorthand to get say `bigint(20)` when doing `->column('foo', 'integer', ['limit' => MysqlAdapter::INT_BIG]`, could just arbitrarily cap them to be values less than `INT_MAX`, and then can declare the `limit` clause as an integer again.

Looks like this is implemented differently than `length` in cakephp (which is just reflecting the mysql metadata). This should probably be implemented differently, but that's a separate issue.

I leave it up to you. I was just reviewing the type conventions and noticed the type changed in an odd way there.

File: tests/Phinx/Db/Adapter/MysqlAdapterTest.php
Comment (MasterOdin) [reply tests/Phinx/Db/Adapter/MysqlAdapterTest.php]: Changed back to `limit` being an integer.

File: src/Phinx/Db/Adapter/AdapterInterface.php
Comment (MasterOdin) [reply src/Phinx/Db/Adapter/AdapterInterface.php]: This was reverted, such that this function expects an integer again.

**Comment by MasterOdin** [comment]
@othercorey @dereuromark I followed through with what I talked about above, giving the `MysqlAdapter` constants values pinned at `PHP_INT_MAX` so that `setLimit` can be an integer again.

> Looks like this is implemented differently than length in cakephp (which is just reflecting the mysql metadata). This should probably be implemented differently, but that's a separate issue.

That's the expected usage of `limit` (or its alias `length`). The constants just exist as a way of specifying a different type for `integer`, `string`, and `blob` base types on mysql. When actually fetching columns from the DB, it's expected that their actual limit / length will be an integer well below `PHP_INT_MAX` as expected.

**Pull request merged**
